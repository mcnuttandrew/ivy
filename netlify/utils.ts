import {MongoClient} from 'mongodb';
import type {Handler} from '@netlify/functions';

const DB_URL = process.env.DB_URL || 'mongodb://localhost:27017';
const DB_NAME = 'ivy-be';

export function getParametersFromPath(path) {
  console.log(path)
  const [_, __, ...parameters] = path.split('/').map((x) => x.replace(/%20/g, ' '));
  const [_0, _1, author, name, ...rest] = parameters;
  return {author, name, instance: rest.join('/')};
}

export function errorResponse(callback, err, statusCode = 500) {
  console.error(err);

  callback(null, {statusCode, body: JSON.stringify({error: err})});
}

export const getMany: (colName: string) => Handler = (colName: string) => (event, context, callback) => {
  MongoClient.connect(`${DB_URL}/${DB_NAME}`)
    .then((connection) =>
      connection
        .db(DB_NAME)
        .collection(colName)
        .find({})
        .toArray()
        .then((results) => {
          callback!(null, {
            statusCode: 200,
            body: JSON.stringify(results),
          });
          connection.close();
        }),
    )
    .catch((err) => errorResponse(callback, err));
};

export const getOne: (
  colName: string,
  pathToQuery: (path: string) => Record<string, string>,
  resultToOutput: (x: any) => any,
) => Handler = (colName, pathToQuery, resultToOutput) => (event, context, callback) => {
  let query;
  try {
    query = pathToQuery(event.path);
  } catch (e) {
    errorResponse(callback, 'Bad submit');
    return;
  }

  MongoClient.connect(`${DB_URL}/${DB_NAME}`)
    .then((connection) => {
      const db = connection.db(DB_NAME);
      const collection = db.collection(colName);

      collection
        .find(query)
        .limit(1)
        .sort({$natural: -1})
        .toArray()
        .then((result) => {
          const output = resultToOutput(result[0]);
          if (output) {
            callback!(null, output);
          } else {
            errorResponse(callback, 'Not found');
          }
          connection.close();
        });
    })
    .catch((err) => errorResponse(callback, err));
};

// should be formatted like 2020-09-09T04:35:08.804Z
const makeDateString = () => new Date().toISOString();

export const insertOne: (colName: string, bodyToQuery: (body: string) => Record<string, any>) => Handler =
  (colName, bodyToQuery) => (event, context, callback) => {
    let query: Record<string, any>;
    try {
      query = bodyToQuery(event.body || '');
    } catch (e) {
      errorResponse(callback, 'Bad submit');
      return;
    }

    MongoClient.connect(`${DB_URL}/${DB_NAME}`)
      .then(async (connection) => {
        const db = connection.db(DB_NAME);
        const collection = db.collection(colName);

        const latestInsert = await collection.find({}).sort({id: -1}).limit(1).toArray();
        const lastId = latestInsert[0]?.id || 0;

        return collection.insertOne({...query, id: lastId + 1, created_at: makeDateString()}).then(() => {
          callback!(null, {statusCode: 200});
          connection.close();
        });
      })
      .catch((err) => errorResponse(callback, err));
  };
