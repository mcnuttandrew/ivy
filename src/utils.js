export function classnames(classObject) {
  return Object.keys(classObject)
    .filter(name => classObject[name])
    .join(" ");
}
