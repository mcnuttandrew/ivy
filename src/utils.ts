export function classnames(classObject: {[val: string]: boolean}): string {
  return Object.keys(classObject)
    .filter(name => classObject[name])
    .join(' ');
}

export function getUniques(data: any, field: string): string[] {
  return Object.keys(
    data.reduce((acc: {[a: string]: boolean}, row: any) => {
      acc[row[field]] = true;
      return acc;
    }, {}),
  );
}

export function getDomain(data: any, field: string): number[] {
  type domainType = {min: number, max: number};
  return Object.values(
    data.reduce(
      (acc: domainType, row: any) => {
        return {
          min: Math.min(acc.min, row[field]),
          max: Math.max(acc.max, row[field]),
        };
      },
      {min: Infinity, max: -Infinity},
    ),
  );
}
