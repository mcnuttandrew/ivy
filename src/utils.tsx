import {List} from 'immutable';
import {TemplateWidget} from './constants/templates';
import {AppState} from './reducers/default-state';
import {DataType, ColumnHeader} from './types';

export function classnames(classObject: {[val: string]: boolean}): string {
  return Object.keys(classObject)
    .filter(name => classObject[name])
    .join(' ');
}

export function getUniques(data: any, field: string): string[] {
  return Object.keys(
    data.reduce((acc: {[a: string]: boolean}, row: any) => {
      acc[row[field]] = true;
      return acc;
    }, {}),
  ).sort();
}

export function getDomain(data: any, field: string): number[] {
  type domainType = {min: number; max: number};
  return Object.values(
    data.reduce(
      (acc: domainType, row: any) => {
        return {
          min: Math.min(acc.min, row[field]),
          max: Math.max(acc.max, row[field]),
        };
      },
      {min: Infinity, max: -Infinity},
    ),
  );
}

export function executePromisesInSeries(tasks: any): any {
  return tasks.reduce(
    (promiseChain: any, task: any): any => promiseChain.then(task),
    Promise.resolve([]),
  );
}

export function findField(
  state: AppState,
  targetField: string,
  columnKey = 'columns',
) {
  return state
    .get(columnKey)
    .find(({field}: {field: string}) => field === targetField);
}

export function compareObjects(a: any, b: any) {
  return JSON.stringify(a) === JSON.stringify(b);
}

const DEFAULT_CONFIG = {
  facet: {width: 150, height: 150},
  overlay: {line: true},
  scale: {useRawDomain: false},
};

export function cleanSpec(spec: any) {
  if (spec.spec) {
    return {
      padding: 100,
      ...spec,
      spec: {
        config: DEFAULT_CONFIG,
        height: 300,
        width: 300,
        ...spec.spec,
      },
      data: {
        name: 'myData',
      },
    };
  }
  return {
    config: DEFAULT_CONFIG,
    padding: 50,
    ...spec,
    data: {
      name: 'myData',
    },
  };
  // return {
  //   config: DEFAULT_CONFIG,
  //   padding: 50,
  //   ...spec,
  //   encoding: {
  //     ...Object.entries(spec.encoding).reduce((acc: any, [key, val]: any) => {
  //       if (!compareObjects(val, {})) {
  //         acc[key] = val;
  //       }
  //       return acc;
  //     }, {}),
  //   },
  // };
}

// safely access elements on a nested object
export function get(obj: any, route: string[]): any {
  if (!obj) {
    return null;
  }
  if (route.length === 0) {
    return null;
  }
  if (route.length === 1) {
    return obj[route[0]];
  }
  const next = obj[route[0]];
  if (!next) {
    return null;
  }
  return get(next, route.slice(1));
}

export function getAllInUseFields(spec: any): Set<string> {
  const inUse = new Set([]);
  const encoding =
    spec.getIn(['spec', 'encoding']) || spec.getIn(['encoding']) || [];
  encoding.forEach((x: any) => {
    if (!x || !x.size) {
      return;
    }
    const channel = x.toJS();
    if (typeof channel.field === 'string') {
      inUse.add(channel.field);
      return;
    }
    if (channel.field && channel.field.repeat) {
      inUse.add(channel.field.repeat);
      return;
    }
  });
  return inUse;
}

export const extractFieldStringsForType = (
  columns: ColumnHeader[],
  type: DataType,
) =>
  columns
    .filter((column: ColumnHeader) => column.type === type)
    .map((column: ColumnHeader) => column.field);

export const checkEncodingForValidity = (spec: any) => {
  const usingNested = !!spec.spec;
  if (usingNested ? spec.spec.layer : spec.layer) {
    console.log('layer');
    return false;
  }
  if (usingNested ? !spec.spec.encoding : !spec.encoding) {
    console.log('encoding');
    return false;
  }
  return true;
};

export const getTemplate = (state: AppState, template: string) => {
  return state.get('templates').find((d: any) => d.templateName === template);
};

export const currentTemplate = (state: AppState) =>
  getTemplate(state, state.get('encodingMode'));

export function widgetInUse(code: string, widgetName: string) {
  return code.match(new RegExp(`\\[${widgetName}\\]`, 'g'));
}
export function allWidgetsInUse(code: string, widgets: List<TemplateWidget>) {
  return widgets
    .filter((widget: TemplateWidget) => widget.widgetType !== 'Text')
    .every((widget: TemplateWidget) => !!widgetInUse(code, widget.widgetName));
}

export const toSelectFormat = (
  arr: string[],
): {value: string; label: string}[] =>
  arr.map((x: string) => ({value: x, label: x}));

// setting dimensions requires that dimension name be wrapped in a string
// here we strip them off so that the channel cencoding can find the correct value
export function trim(dimName: string) {
  if (!dimName || dimName.length < 2) {
    return dimName;
  }
  if (dimName[0] === '"' && dimName[dimName.length - 1] === '"') {
    return dimName.slice(1, dimName.length - 1);
  }
  return dimName;
}
