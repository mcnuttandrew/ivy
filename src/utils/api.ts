import {Template} from '../types';
import {DEFAULT_TEMPLATES} from '../templates';
import {FETCH_PARMS} from '../constants/index';

function simpleGetJSON(url: string): Promise<any> {
  return fetch(url, FETCH_PARMS as any)
    .then((x) => x.json())
    .catch((e) => console.error(e));
}

export function getTemplates(): Promise<Template[]> {
  return fetch(`.netlify/functions/templates`, FETCH_PARMS as any)
    .then((x) => x.json())
    .then((fetchedData) =>
      fetchedData.reduce((acc: any[], x: any) => {
        try {
          let result;
          if (typeof x.template === 'string') {
            result = JSON.parse(x.template);
          } else if (typeof x.template === 'object') {
            result = x.template;
          } else {
            throw new Error('bad template');
          }
          acc.push(result);
        } catch (e) {
          console.error('parse fail !', x);
        }
        return acc;
      }, []),
    );
}

export function getTemplate(templateAuthor: string, templateName: string): Promise<Template> {
  return new Promise((resolve, reject) => {
    const foundTemplate = DEFAULT_TEMPLATES.find(
      (template) => template.templateAuthor === templateAuthor && template.templateName === templateName,
    );
    if (foundTemplate) {
      resolve(foundTemplate);
      return;
    }
    return simpleGetJSON(`.netlify/functions/template/${templateAuthor}/${templateName}`)
      .then((template) => resolve(template))
      .catch((e) => reject(e));
  });
}

interface TemplateInstance {
  created_at: string;
  dataset: string;
  id: number;
  instance_creator?: string;
  name: string;
  template_creator: string;
  template_instance: {[key: string]: string | string[]};
  template_name: string;
}

export function getTemplateInstance(
  templateAuthor: string,
  templateName: string,
  templateInstance: string,
): Promise<TemplateInstance> {
  return simpleGetJSON(
    `.netlify/functions/template-instance/${templateAuthor}/${templateName}/${templateInstance}`,
  );
}
