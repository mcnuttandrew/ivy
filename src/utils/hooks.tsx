import {useEffect, useState} from 'react';

type windowSize = {width: number; height: number};
export function useWindowSize(): windowSize {
  const isClient = typeof window === 'object';

  function getSize(): windowSize {
    return {
      width: isClient ? window.innerWidth : undefined,
      height: isClient ? window.innerHeight : undefined,
    };
  }

  const [windowSize, setWindowSize] = useState(getSize);

  useEffect(() => {
    if (!isClient) {
      return;
    }

    function handleResize(): void {
      setWindowSize(getSize());
    }

    window.addEventListener('resize', handleResize);
    return (): any => window.removeEventListener('resize', handleResize);
  }, []); // Empty array ensures that effect is only run on mount and unmount

  return windowSize;
}
