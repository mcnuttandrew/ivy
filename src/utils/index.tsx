import stringify from 'json-stringify-pretty-compact';
import {
  DataTargetWidget,
  DataType,
  MultiDataTargetWidget,
  Template,
  TemplateMap,
  TemplateWidget,
  WidgetSubType,
} from '../templates/types';
import {AppState} from '../reducers/default-state';
import {TEXT_TYPE} from '../constants/index';
import NONE from '../templates/example-templates/none';
import {ColumnHeader} from '../types';

/* eslint-disable @typescript-eslint/no-empty-function*/
export const NULL = (): void => {};
/* eslint-enable @typescript-eslint/no-empty-function*/

export function classnames(classObject: {[val: string]: boolean}): string {
  return Object.keys(classObject)
    .filter(name => classObject[name] && name)
    .join(' ');
}

export function getUniques(data: any, field: string): string[] {
  return Object.keys(
    data.reduce((acc: {[a: string]: boolean}, row: any) => {
      acc[row[field]] = true;
      return acc;
    }, {}),
  ).sort();
}

export function getDomain(data: any, field: string): number[] {
  type domainType = {min: number; max: number};
  return Object.values(
    data.reduce(
      (acc: domainType, row: any) => {
        return {
          min: Math.min(acc.min, row[field]),
          max: Math.max(acc.max, row[field]),
        };
      },
      {min: Infinity, max: -Infinity},
    ),
  );
}

export function executePromisesInSeries(tasks: any): any {
  return tasks.reduce((promiseChain: any, task: any): any => promiseChain.then(task), Promise.resolve([]));
}

export function findField(state: AppState, targetField: string, columnKey = 'columns'): ColumnHeader {
  if (columnKey === 'metaColumns') {
    return state.metaColumns.find(({field}: {field: string}) => field === targetField);
  }
  return state.columns.find(({field}: {field: string}) => field === targetField);
}

// safely access elements on a nested object
export function get(obj: any, route: (string | number)[]): any {
  if (!obj) {
    return null;
  }
  if (route.length === 0) {
    return null;
  }
  if (route.length === 1) {
    return obj[route[0]];
  }
  const next = obj[route[0]];
  if (!next) {
    return null;
  }
  return get(next, route.slice(1));
}

export function getAllInUseFields(spec: any): Set<string> {
  // this only works for vega-lite
  const inUse = new Set([]);
  const encoding = (spec.spec && spec.spec.encoding) || spec.encoding || {};
  Object.values(encoding).forEach((x: any) => {
    if (!x) {
      return;
    }
    const channel = x;
    if (typeof channel.field === 'string') {
      inUse.add(channel.field);
      return;
    }
    if (channel.field && channel.field.repeat) {
      inUse.add(channel.field.repeat);
      return;
    }
  });
  return inUse;
}

export const extractFieldStringsForType = (columns: ColumnHeader[], type: DataType): string[] =>
  columns.filter((column: ColumnHeader) => column.type === type).map((column: ColumnHeader) => column.field);

export const getTemplate = (state: AppState, template: string): Template | null => {
  return state.templates.find((d: any) => d.templateName === template);
};

export function widgetInUse(code: string, name: string): boolean {
  return Boolean(code.match(new RegExp(`\\[${name}\\]`, 'g')));
}
export function allWidgetsInUse(code: string, widgets: TemplateWidget<WidgetSubType>[]): boolean {
  return widgets
    .filter((widget: TemplateWidget<WidgetSubType>) => widget.type !== 'Text')
    .every((widget: TemplateWidget<WidgetSubType>) => !!widgetInUse(code, widget.name));
}

export const toSelectFormat = (arr: string[]): {value: string; label: string}[] =>
  arr.map((x: string) => ({value: x, label: x}));

// setting dimensions requires that dimension name be wrapped in a string
// here we strip them off so that the channel cencoding can find the correct value
export function trim(dimName: string): string {
  if (!dimName || dimName.length < 2) {
    return dimName;
  }
  if (dimName[0] === '"' && dimName[dimName.length - 1] === '"') {
    return dimName.slice(1, dimName.length - 1);
  }
  return dimName;
}

export const toList = (list: string[]): {display: string; value: string}[] =>
  list.map(display => ({
    display,
    value: `"${display}"`,
  }));

export function serializeTemplate(template: Template): string {
  return stringify({
    $schema: 'https://kind-goldwasser-f3ce26.netlify.com/assets/hydra-template.json',
    templateName: template.templateName,
    templateDescription: template.templateDescription,
    templateAuthor: template.templateAuthor,
    code: 'SEE BODY',
    templateLanguage: template.templateLanguage,
    widgets: template.widgets,
  });
}

export function deserializeTemplate(templateString: string): Template {
  const code = JSON.parse(templateString);
  return {
    templateName: code.templateName,
    templateDescription: code.templateDescription,
    code: 'SEE CODE EDITOR',
    templateLanguage: code.templateLanguage,
    widgets: code.widgets,
    templateAuthor: code.templateAuthor,
  };
}

type SaveState = 'NA' | 'NOT FOUND' | 'EQUAL' | 'DIFFERENT';
export function getTemplateSaveState(base: AppState): SaveState {
  const template = base.currentTemplateInstance;
  // using the grammar mode
  if (!template) {
    return 'NA';
  }
  const associatedUpstreamTemplate = getTemplate(base, template.templateName);
  if (!associatedUpstreamTemplate) {
    return 'NOT FOUND';
  }
  return associatedUpstreamTemplate === template ? 'EQUAL' : 'DIFFERENT';
}

const USE_LOCAL = false;
export function serverPrefix(): string {
  return USE_LOCAL ? 'http://localhost:5000' : 'https://hydra-template-server.herokuapp.com';
}

export const computeValidAddNexts = (template: Template, templateMap: TemplateMap): Set<string> => {
  const dims = ['DIMENSION', 'MEASURE', 'METACOLUMN', 'TIME'];
  const dimCounter = dims.reduce((acc: any, key) => ({...acc, [key]: []}), {});

  const toSet = (counter: {[x: string]: any[]}): Set<string> =>
    new Set(
      Object.entries(counter)
        .filter(x => x[1].length)
        .map(x => x[0]),
    );

  if (!template) {
    dims.forEach(x => dimCounter[x].push(x));
    // don't do anything with T0 for now
    return toSet(dimCounter);
  }
  const result = template.widgets
    .filter(d => ['MultiDataTarget', 'DataTarget'].includes(d.type))
    .reduce((acc: any, widget: any) => {
      const type = widget.type;
      const name = widget.name;
      const val = templateMap[name];

      if (
        type === 'MultiDataTarget' &&
        (!val || val.length < widget.config.maxNumberOfTargets || !widget.config.maxNumberOfTargets)
      ) {
        widget.config.allowedTypes.forEach((allowedType: string): void => acc[allowedType].push(name));
      }
      // dont try figure it out if it's in use, needs to be before multidatatarget which has a truthy null, []
      if (val) {
        return acc;
      }

      if (type === 'DataTarget') {
        widget.config.allowedTypes.forEach((allowedType: string): void => acc[allowedType].push(name));
      }
      return acc;
    }, dimCounter);
  return toSet(result);
};

export const makeColNameMap = (columns: ColumnHeader[]): {[d: string]: ColumnHeader} =>
  columns.reduce((acc: {[d: string]: ColumnHeader}, colKey: ColumnHeader) => {
    acc[colKey.field] = colKey;
    return acc;
  }, {});

export function buildCounts(template: Template): any {
  let SUM = 0;
  const counts = template.widgets.reduce(
    (acc: any, row: TemplateWidget<WidgetSubType>) => {
      if (row.type === 'DataTarget') {
        const {allowedTypes, required} = row.config as DataTargetWidget;
        if (!required) {
          return acc;
        }
        SUM += 1;

        allowedTypes.forEach((type: string) => {
          acc[type] += 1;
          if (allowedTypes.length > 1) {
            acc.mixingOn = acc.mixingOn.add(type);
          }
        });
      }
      if (row.type === 'MultiDataTarget') {
        const {allowedTypes, minNumberOfTargets, required} = row.config as MultiDataTargetWidget;
        if (!required) {
          return acc;
        }
        SUM += Number(minNumberOfTargets) || 0;
        allowedTypes.forEach((type: string) => {
          acc[type] += Number(minNumberOfTargets) || 0;
          if (allowedTypes.length > 1) {
            acc.mixingOn = acc.mixingOn.add(type);
          }
        });
      }
      return acc;
    },
    {DIMENSION: 0, MEASURE: 0, TIME: 0, mixingOn: new Set()},
  );
  return {DIMENSION: counts.DIMENSION, MEASURE: counts.MEASURE, TIME: counts.TIME, SUM};
}

export function searchDimensionsCanMatch(
  template: Template,
  targetCols: string[],
  columns: ColumnHeader[],
): {canBeUsed: boolean; isComplete: boolean} {
  const colMap = makeColNameMap(columns);
  const desiredColumns = targetCols.map(key => colMap[trim(key)]);
  const config = template.widgets;
  const targets = config.filter(d => d.type === 'DataTarget') as TemplateWidget<DataTargetWidget>[];
  const multiTargets = config.filter(d => d.type === 'MultiDataTarget') as TemplateWidget<
    MultiDataTargetWidget
  >[];
  const numRequired = targets.filter(d => d.config.required).length;
  const usedTargets: Set<string> = new Set([]);
  const result = desiredColumns.every(col => {
    const availableSingleTargetField = targets
      .filter(d => !usedTargets.has(d.name))
      .find(d => col && d.config.allowedTypes.includes(col.type));
    const availableMultiTargetField = multiTargets.find(d => col && d.config.allowedTypes.includes(col.type));

    if (availableSingleTargetField) {
      usedTargets.add(availableSingleTargetField.name);
      return true;
    } else if (availableMultiTargetField) {
      return true;
    } else {
      return false;
    }
  });
  return {canBeUsed: result, isComplete: usedTargets.size >= numRequired};
}

// used for filtering out unsearched templates
export function searchPredicate(
  searchKey: string,
  templateName: string,
  templateDescription: string,
): boolean {
  const matchDescription = templateDescription && templateDescription.toLowerCase().includes(searchKey || '');
  const matchName = templateName && templateName.toLowerCase().includes(searchKey || '');
  return matchDescription || matchName;
}

export function sortObjectAlphabetically(obj: any): any {
  return Object.entries(obj)
    .sort((a, b) => a[0].localeCompare(b[0]))
    .reduce((acc: any, [key, val]) => {
      acc[key] = val;
      return acc;
    }, {});
}

export function getTemplateName(template: Template | null): string {
  if (!template) {
    return 'T0';
  }
  return template && template.templateName === NONE.templateName ? 'Template Gallery' : template.templateName;
}

export function union(setA: Set<any>, setB: Set<any>): Set<any> {
  const _union = new Set(setA);
  for (const elem of setB) {
    _union.add(elem);
  }
  return _union;
}
export function difference(setA: Set<any>, setB: Set<any>): Set<any> {
  const _difference = new Set(setA);
  for (const elem of setB) {
    _difference.delete(elem);
  }
  return _difference;
}

export function safeParse(code: string): string | boolean {
  let x = null;
  try {
    x = JSON.parse(code);
  } catch (e) {
    x = false;
  }
  return x;
}

export function makeCustomType(field: string): ColumnHeader {
  return {type: 'CUSTOM', field, originalType: 'CUSTOM', domain: []};
}

interface MakeOptionsForDropdownProps {
  template: Template;
  columns: ColumnHeader[];
  widget: TemplateWidget<DataTargetWidget | MultiDataTargetWidget>;
  useGroupsAsTypes?: boolean;
}
export function makeOptionsForDropdown(
  props: MakeOptionsForDropdownProps,
): {display: string; value: string; group: string | null}[] {
  const {template, columns, widget, useGroupsAsTypes} = props;
  return [
    {display: 'Select a value', value: null, group: null},
    ...(template.customCards || []).map(card => ({display: card, value: card, group: 'Template Fields'})),
  ].concat(
    columns
      .map(column => ({
        display: `${column.field} ${TEXT_TYPE[column.type]}`,
        value: column.field,
        group: useGroupsAsTypes
          ? column.type
          : widget.config.allowedTypes.includes(column.type)
          ? 'RECOMENDED'
          : 'OUT OF TYPE',
      }))
      .sort((a, b) => a.display.localeCompare(b.display)),
  );
}
