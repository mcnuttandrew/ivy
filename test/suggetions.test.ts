import {IsotypeBarChart} from './text-examples';
import {synthesizeSuggestions, takeSuggestion} from '../src/utils/introspect';

test('#synthesizeSuggestions', () => {
  const suggestions = synthesizeSuggestions(IsotypeBarChart, []);
  expect(suggestions).toMatchSnapshot();
  // remove the foreign data
  const newCode = takeSuggestion(IsotypeBarChart, suggestions[3]);
  const suggestions2 = synthesizeSuggestions(newCode, []);
  expect(suggestions2).toMatchSnapshot();

  const newCod2 = takeSuggestion(newCode, suggestions2[2]);
  const widget = suggestions2[2].sideEffect();
  const widgets = [widget];
  const suggestions3 = synthesizeSuggestions(newCod2, widgets);
  expect(suggestions3).toMatchSnapshot();
});
