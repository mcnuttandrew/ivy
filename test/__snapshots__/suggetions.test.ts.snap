// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#synthesizeSuggestions 1`] = `
Array [
  Object {
    "comment": "col -> Dim1 (CREATE Dim1)",
    "from": "\\"col\\"",
    "sideEffect": [Function],
    "simpleReplace": false,
    "to": "\\"[Dim1]\\"",
  },
  Object {
    "comment": "animal -> Dim1 (CREATE Dim1)",
    "from": "\\"animal\\"",
    "sideEffect": [Function],
    "simpleReplace": false,
    "to": "\\"[Dim1]\\"",
  },
  Object {
    "comment": "country -> Dim1 (CREATE Dim1)",
    "from": "\\"country\\"",
    "sideEffect": [Function],
    "simpleReplace": false,
    "to": "\\"[Dim1]\\"",
  },
  Object {
    "comment": "remove specific data",
    "from": "\\"values\\": [
      {\\"country\\": \\"Great Britain\\", \\"animal\\": \\"cattle\\", \\"col\\": 3},
      {\\"country\\": \\"Great Britain\\", \\"animal\\": \\"cattle\\", \\"col\\": 2},
      {\\"country\\": \\"Great Britain\\", \\"animal\\": \\"cattle\\", \\"col\\": 1},
      {\\"country\\": \\"Great Britain\\", \\"animal\\": \\"pigs\\", \\"col\\": 2},
      {\\"country\\": \\"Great Britain\\", \\"animal\\": \\"pigs\\", \\"col\\": 1},
      {\\"country\\": \\"Great Britain\\", \\"animal\\": \\"sheep\\", \\"col\\": 10},
      {\\"country\\": \\"Great Britain\\", \\"animal\\": \\"sheep\\", \\"col\\": 9},
      {\\"country\\": \\"Great Britain\\", \\"animal\\": \\"sheep\\", \\"col\\": 8},
      {\\"country\\": \\"Great Britain\\", \\"animal\\": \\"sheep\\", \\"col\\": 7},
      {\\"country\\": \\"Great Britain\\", \\"animal\\": \\"sheep\\", \\"col\\": 6},
      {\\"country\\": \\"Great Britain\\", \\"animal\\": \\"sheep\\", \\"col\\": 5},
      {\\"country\\": \\"Great Britain\\", \\"animal\\": \\"sheep\\", \\"col\\": 4},
      {\\"country\\": \\"Great Britain\\", \\"animal\\": \\"sheep\\", \\"col\\": 3},
      {\\"country\\": \\"Great Britain\\", \\"animal\\": \\"sheep\\", \\"col\\": 2},
      {\\"country\\": \\"Great Britain\\", \\"animal\\": \\"sheep\\", \\"col\\": 1},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"cattle\\", \\"col\\": 9},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"cattle\\", \\"col\\": 8},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"cattle\\", \\"col\\": 7},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"cattle\\", \\"col\\": 6},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"cattle\\", \\"col\\": 5},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"cattle\\", \\"col\\": 4},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"cattle\\", \\"col\\": 3},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"cattle\\", \\"col\\": 2},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"cattle\\", \\"col\\": 1},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"pigs\\", \\"col\\": 6},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"pigs\\", \\"col\\": 5},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"pigs\\", \\"col\\": 4},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"pigs\\", \\"col\\": 3},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"pigs\\", \\"col\\": 2},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"pigs\\", \\"col\\": 1},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"sheep\\", \\"col\\": 7},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"sheep\\", \\"col\\": 6},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"sheep\\", \\"col\\": 5},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"sheep\\", \\"col\\": 4},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"sheep\\", \\"col\\": 3},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"sheep\\", \\"col\\": 2},
      {\\"country\\": \\"United States\\", \\"animal\\": \\"sheep\\", \\"col\\": 1}
    ]",
    "simpleReplace": true,
    "to": "\\"name\\": \\"myData\\"",
  },
]
`;

exports[`#synthesizeSuggestions 2`] = `
Array [
  Object {
    "comment": "col -> Dim1 (CREATE Dim1)",
    "from": "\\"col\\"",
    "sideEffect": [Function],
    "simpleReplace": false,
    "to": "\\"[Dim1]\\"",
  },
  Object {
    "comment": "animal -> Dim1 (CREATE Dim1)",
    "from": "\\"animal\\"",
    "sideEffect": [Function],
    "simpleReplace": false,
    "to": "\\"[Dim1]\\"",
  },
  Object {
    "comment": "country -> Dim1 (CREATE Dim1)",
    "from": "\\"country\\"",
    "sideEffect": [Function],
    "simpleReplace": false,
    "to": "\\"[Dim1]\\"",
  },
]
`;

exports[`#synthesizeSuggestions 3`] = `
Array [
  Object {
    "comment": "col -> Dim1",
    "from": "\\"col\\"",
    "simpleReplace": false,
    "to": "\\"[Dim1]\\"",
  },
  Object {
    "comment": "col -> Dim2 (CREATE Dim2)",
    "from": "\\"col\\"",
    "sideEffect": [Function],
    "simpleReplace": false,
    "to": "\\"[Dim2]\\"",
  },
  Object {
    "comment": "animal -> Dim1",
    "from": "\\"animal\\"",
    "simpleReplace": false,
    "to": "\\"[Dim1]\\"",
  },
  Object {
    "comment": "animal -> Dim2 (CREATE Dim2)",
    "from": "\\"animal\\"",
    "sideEffect": [Function],
    "simpleReplace": false,
    "to": "\\"[Dim2]\\"",
  },
]
`;
